tests/unit/aws/test_proxy.py

```diff
@@ -35,3 +37,55 @@ def list_queues(self, context, request):
             assert result["QueueUrls"] == [
                 "http://localhost:4566/000000000000/foo-queue",
             ]
+
+
+class TestAsfWithFallbackListener:
+    def test_request_response(self):
+        provider_calls = []
+        fallback_calls = []
+
```

Comment:
I can see this setup being widely reused, the pair `Provider`/`Fallback` would be interesting to have fixtures to pull them into different tests and not pollute the setup.

Reply:
yep good point. it would probably be good to use mocks in this case that already record calls internally.